src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1324,6 +1329,31 @@ object Desugar {
 
           case PNegation(op) => for {o <- go(op)} yield in.Negation(o)(src)
 
+          case PEquals(left, right) if info.typOfExprOrType(left) == PermissionT || info.typOfExprOrType(right) == PermissionT =>
+            violation(left.isInstanceOf[PExpression], s"Expected an expression but got $left instead.")
+            violation(right.isInstanceOf[PExpression], s"Expected an expression but got $right instead.")
+            for {
+              l <- permissionD(ctx)(left.asInstanceOf[PExpression])
+              r <- permissionD(ctx)(right.asInstanceOf[PExpression])
+            } yield in.EqCmp(l, r)(src)
+
+          case PUnequals(left, right) if info.typOfExprOrType(left) == PermissionT || info.typOfExprOrType(right) == PermissionT =>
+            violation(left.isInstanceOf[PExpression], s"Expected an expression but got $left instead.")
+            violation(right.isInstanceOf[PExpression], s"Expected an expression but got $right instead.")
+            for {
+              l <- permissionD(ctx)(left.asInstanceOf[PExpression])
+              r <- permissionD(ctx)(right.asInstanceOf[PExpression])
+            } yield in.UneqCmp(l, r)(src)
+
+          case PLess(left, right) if info.typ(left) == PermissionT || info.typ(right) == PermissionT =>
```

Comment:
I think it would be better to make the distinction only once at the desugaring of PLess and not once for every possibility.